### PR TITLE
Proper handling of the empty "binding:profile"

### LINKED ISSUE #756: Port binding support of PolicyNotAuthorized
Hi there,

Currently I encountered a port bonding issue with openstack_networking_port_v2.

### Terraform Version
Terraform v0.11.13
terraform-provider-openstack_v1.17.0_x4

### Affected Resource(s) and Terraform Configuration Files

resource "openstack_networking_port_v2" "port_net-02_v4" {
    name = "${var.name_prefix}-${var.hostname}"
    network_id = "${var.net-02_net_uuid}"
    admin_state_up = "true"

    binding = {
        vnic_type = "direct"
    }
    security_group_ids = [
  ]
    fixed_ip {
        "subnet_id" = "${var.net-02_v4_sub_uuid}"
    }
    count = "${var.count}"
}

### Debug Output

2019-05-22 01:34:20,216 DEBUG: * module.eason-worker-01.openstack_networking_port_v2.port_net-02_v4: 1 error(s) occurred:
2019-05-22 01:34:20,216 DEBUG: 
2019-05-22 01:34:20,217 DEBUG: * openstack_networking_port_v2.port_net-02_v4: Error creating openstack_networking_port_v2: Request forbidden: [POST https://10.4.166.166:13696/v2.0/ports], error message: {"NeutronError": {"message": "(((rule:create_port and rule:create_port:binding:profile) and rule:create_port:fixed_ips) and rule:create_port:binding:vnic_type) on {'binding:host_id': <neutron_lib.constants.Sentinel object at 0x345ffd0>, u'name': u'eason-worker-01', 'allowed_address_pairs': <neutron_lib.constants.Sentinel object at 0x345ffd0>, u'admin_state_up': True, u'network_id': u'**7113fc5d-a805-437c-ac2e-442acc6e7b82**', 'tenant_id': u'498c8a0646e34dccbe2fd598e9059f95', 'extra_dhcp_opts': None, 'mac_address': <neutron_lib.constants.Sentinel object at 0x345ffd0>, u'binding:vnic_type': u'direct', 'device_owner': '', 'qos_policy_id': None, 'device_id': '', 'port_security_enabled': <neutron_lib.constants.Sentinel object at 0x345ffd0>, u'binding:profile': {}, 'project_id': u'498c8a0646e34dccbe2fd598e9059f95', u'fixed_ips': [{u'subnet_id': u'**316c3c7a-fa99-45d0-ae00-fb31e1372734**'}], 'security_groups': <neutron_lib.constants.Sentinel object at 0x345ffd0>, 'description': ''} by {'domain': None, 'project_name': u'xxxx', 'tenant_name': u'xxxx', 'project_domain': None, 'timestamp': '2019-05-22 01:34:18.897611', 'auth_token': 'db015113f5a247989ca71d81e70203ab', 'resource_uuid': None, 'is_admin': False, 'user': u'7aea1ee662874a29b9d0a11ae2303df5', 'tenant': u'498c8a0646e34dccbe2fd598e9059f95', 'read_only': False, 'project_id': u'498c8a0646e34dccbe2fd598e9059f95', 'user_id': u'7aea1ee662874a29b9d0a11ae2303df5', 'show_deleted': False, 'roles': [u'_member_'], 'user_identity': u'7aea1ee662874a29b9d0a11ae2303df5 498c8a0646e34dccbe2fd598e9059f95 - - -', 'tenant_id': u'498c8a0646e34dccbe2fd598e9059f95', 'is_admin_project': True, 'request_id': 'req-73646982-9625-4f09-8796-6b63628e2bbb', 'user_domain': None, 'user_name': u'xxxx'} disallowed by policy", "type": "PolicyNotAuthorized", "detail": ""}}

2019-05-22 01:34:14,068 DEBUG: module.eason-worker-01.openstack_networking_port_v2.port_net-02_v4: Creating...
2019-05-22 01:34:14,068 DEBUG:   admin_state_up:           "" => "true"
2019-05-22 01:34:14,068 DEBUG:   all_fixed_ips.#:          "" => "<computed>"
2019-05-22 01:34:14,068 DEBUG:   all_security_group_ids.#: "" => "<computed>"
2019-05-22 01:34:14,068 DEBUG:   all_tags.#:               "" => "<computed>"
2019-05-22 01:34:14,069 DEBUG:   binding.#:                "" => "1"
2019-05-22 01:34:14,069 DEBUG:   binding.0.vif_details.%:  "" => "<computed>"
2019-05-22 01:34:14,069 DEBUG:   binding.0.vif_type:       "" => "<computed>"
2019-05-22 01:34:14,069 DEBUG:   binding.0.vnic_type:      "" => "direct"
2019-05-22 01:34:14,069 DEBUG:   device_id:                "" => "<computed>"
2019-05-22 01:34:14,069 DEBUG:   device_owner:             "" => "<computed>"
2019-05-22 01:34:14,069 DEBUG:   dns_assignment.#:         "" => "<computed>"
2019-05-22 01:34:14,070 DEBUG:   fixed_ip.#:               "" => "1"
2019-05-22 01:34:14,070 DEBUG:   fixed_ip.0.subnet_id:     "" => "316c3c7a-fa99-45d0-ae00-fb31e1372734"
2019-05-22 01:34:14,070 DEBUG:   mac_address:              "" => "<computed>"
2019-05-22 01:34:14,070 DEBUG:   name:                     "" => "eason-worker-01"
2019-05-22 01:34:14,070 DEBUG:   network_id:               "" => "7113fc5d-a805-437c-ac2e-442acc6e7b82"
2019-05-22 01:34:14,070 DEBUG:   port_security_enabled:    "" => "<computed>"
2019-05-22 01:34:14,070 DEBUG:   region:                   "" => "<computed>"
2019-05-22 01:34:14,070 DEBUG:   tenant_id:                "" => "<computed>"

However, when I use openstack CLI, it works:

[eason-dp:/opt/terraform/openstack/resources]# openstack port create --network 7113fc5d-a805-437c-ac2e-442acc6e7b82 --fixed-ip subnet=316c3c7a-fa99-45d0-ae00-fb31e1372734 --vnic-type direct  es-test-sriov-2 
+-------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Field                   | Value                                                                                                                                                                                  |
+-------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| admin_state_up          | UP                                                                                                                                                                                     |
| allowed_address_pairs   |                                                                                                                                                                                        |
| binding_host_id         | None                                                                                                                                                                                   |
| binding_profile         | None                                                                                                                                                                                   |
| binding_vif_details     | None                                                                                                                                                                                   |
| binding_vif_type        | None                                                                                                                                                                                   |
| binding_vnic_type       | direct                                                                                                                                                                                 |
| created_at              | 2019-05-22T02:52:07Z                                                                                                                                                                   |
| data_plane_status       | None                                                                                                                                                                                   |
| description             |                                                                                                                                                                                        |
| device_id               |                                                                                                                                                                                        |
| device_owner            |                                                                                                                                                                                        |
| dns_assignment          | None                                                                                                                                                                                   |
| dns_domain              | None                                                                                                                                                                                   |
| dns_name                | None                                                                                                                                                                                   |
| extra_dhcp_opts         |                                                                                                                                                                                        |
| fixed_ips               | ip_address='10.76.98.75', subnet_id='316c3c7a-fa99-45d0-ae00-fb31e1372734'                                                                                                             |
| id                      | 410650b5-41b5-45c2-83ba-aff0c8519fbb                                                                                                                                                   |
| location                | Munch({'project': Munch({'domain_id': None, 'id': u'498c8a0646e34dccbe2fd598e9059f95', 'name': 'xxxx', 'domain_name': None}), 'cloud': '', 'region_name': '', 'zone': None}) |
| mac_address             | fa:16:3e:9f:dc:64                                                                                                                                                                      |
| name                    | es-test-sriov-2                                                                                                                                                                        |
| network_id              | 7113fc5d-a805-437c-ac2e-442acc6e7b82                                                                                                                                                   |
| port_security_enabled   | True                                                                                                                                                                                   |
| project_id              | 498c8a0646e34dccbe2fd598e9059f95                                                                                                                                                       |
| propagate_uplink_status | None                                                                                                                                                                                   |
| qos_policy_id           | None                                                                                                                                                                                   |
| resource_request        | None                                                                                                                                                                                   |
| revision_number         | 6                                                                                                                                                                                      |
| security_group_ids      | 228fa7f2-a097-413a-9633-085f5de80dce                                                                                                                                                   |
| status                  | DOWN                                                                                                                                                                                   |
| tags                    |                                                                                                                                                                                        |
| trunk_details           | None                                                                                                                                                                                   |
| updated_at              | 2019-05-22T02:52:07Z                                                                                                                                                                   |
+-------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

I just want to make sure it is OpenStack policy issue, then I could raise a ticket to change it by lab admin or my usage of terraform binding block  is incorrect?

Thanks in advance.

### PR DESCRIPTION
**Pull request by kayrus**
Resolves #756

### CODE REVIEW

**Comment by jtopjian** [comment]
recheck

**Comment by kayrus** [comment]
recheck

**Comment by kayrus** [comment]
recheck

File: openstack/resource_openstack_networking_port_v2.go [hunk 1]
```diff
@@ -226,8 +226,8 @@ func resourceNetworkingPortV2() *schema.Resource {
 						"profile": {
 							Type:             schema.TypeString,
 							Optional:         true,
-							ValidateFunc:     validation.ValidateJsonString,
-							DiffSuppressFunc: suppressDiffPortBindingProfileV2,
+							ValidateFunc:     validateJsonObject,
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_port_v2.go]: Why is this being changed?

Comment (kayrus) [reply openstack/resource_openstack_networking_port_v2.go]: * Validation has to allow `null` and doesn't allow `[]`
* diff suppress has to allow `null` as well

Comment (jtopjian) [reply openstack/resource_openstack_networking_port_v2.go]: Ah - understood. Thanks!

Comment (jtopjian) [reply openstack/resource_openstack_networking_port_v2.go]: `validation.ValidateJsonString` seems to still be working for me (for the existing acceptance tests). 

Would you be able to include a test where the response has `null` in it? Unless I'm mistaken, `null` should be interpreted as nil or omited and not an actual string of `"null"`.

**Comment by pkqsun** [comment]
Hi @kayrus  @jtopjian ,

Is any update for this fix ? 
Acturally, I am waiting for the new release/binary with this fix.  🙂

**Comment by kayrus** [comment]
@pkqsun sorry for the delay, I'll try to finish the PR this week. Please note that usually binary releases are not published right after the fix. However you can always compile the binary directly from the source and I suggested to do this in https://github.com/terraform-providers/terraform-provider-openstack/issues/756#issuecomment-494783342

**Comment by kayrus** [comment]
@jtopjian same uppercase IPv6 issue as in the https://github.com/terraform-providers/terraform-provider-openstack/pull/765#issuecomment-499385579

**Comment by jtopjian** [comment]
@kayrus I don't have time right now to look into the cause. You're more than welcome to look into it. It's not a consistent failure and it works on all environments I have access to, so it's hard for me to pinpoint with limited time.

**Comment by kayrus** [comment]
recheck

**Comment by jtopjian** [comment]
If it's consistently failing, it might be a change or bug in the latest OpenStack master/head. I'll have to set up a devstack environment to further test.

**Comment by kayrus** [comment]
@jtopjian I don't think it is an OpenStack issue. Probably it is a terraform issue. Take a look on the log: https://gist.github.com/kayrus/b3993f19a647e135267fed0edf023736

you can see that the request body contains uppercase, when it should be lowercase (according to the `StateFunc` assigned to the `cidr` attribute):

```
2019/06/06 16:35:54 [DEBUG] OpenStack Request Body: {
  "security_group_rule": {
    "cidr": "2001:558:FC00::/39",
    "from_port": 0,
    "ip_protocol": "icmp",
    "parent_group_id": "f76a22c6-1ced-40e7-8b00-c96e1f51a3f4",
    "to_port": 0
  }
}
```

**Comment by jtopjian** [comment]
Having the request sent in upper case is fine.

It's the response that should be lower-case. That's why we use the StateFunc: it takes the user's upper-case input and stores it as lower-case to match the response:

```
2019/06/06 19:25:09 [DEBUG] OpenStack Request Headers:
Accept: application/json
Content-Type: application/json
User-Agent: Terraform/0.12.0 gophercloud/2.0.0
X-Auth-Token: ***
2019/06/06 19:25:09 [DEBUG] OpenStack Request Body: {
  "security_group_rule": {
    "cidr": "2001:558:FC00::/39",
    "from_port": 0,
    "ip_protocol": "icmp",
    "parent_group_id": "5f097b2c-5e1a-41a1-b7dc-56c1e182fbb3",
    "to_port": 0
  }
}
2019/06/06 19:25:10 [DEBUG] OpenStack Response Code: 200
2019/06/06 19:25:10 [DEBUG] OpenStack Response Headers:
Content-Length: 240
Content-Type: application/json
Openstack-Api-Version: compute 2.1
Vary: OpenStack-API-Version
Vary: X-OpenStack-Nova-API-Version
X-Compute-Request-Id: req-0baa7e51-2474-48e4-995e-d30e344164b4
X-Openstack-Nova-Api-Version: 2.1
X-Openstack-Request-Id: req-0baa7e51-2474-48e4-995e-d30e344164b4
2019/06/06 19:25:10 [DEBUG] OpenStack Response Body: {
  "security_group_rule": {
    "from_port": 0,
    "group": {},
    "id": "d87e999b-5b68-44f0-a0ac-5f1801342834",
    "ip_protocol": "icmp",
    "ip_range": {
      "cidr": "2001:558:fc00::/39"
    },
    "parent_group_id": "5f097b2c-5e1a-41a1-b7dc-56c1e182fbb3",
    "to_port": 0
  }
}
```

**Comment by jtopjian** [comment]
I see the issue:

Your results:

```
   "security_group": {
     "description": "first test security group",
     "id": "f76a22c6-1ced-40e7-8b00-c96e1f51a3f4",
     "name": "sg_1",
     "rules": [
       {
         "from_port": 0,
         "group": {},
         "id": "18a96a96-5371-41a9-8c4a-ba16adfceeb5",
         "ip_protocol": "ipv6-icmp",
         "ip_range": {
           "cidr": "2001:558:fc00::/39"
         },
         "parent_group_id": "f76a22c6-1ced-40e7-8b00-c96e1f51a3f4",
         "to_port": 0
       }
     ],
     "tenant_id": "a21b68a7f1bc47babce4d6f54cd75eb9"
   }
 }
```

My output:

```
  "security_group": {
    "description": "first test security group",
    "id": "5f097b2c-5e1a-41a1-b7dc-56c1e182fbb3",
    "name": "sg_1",
    "rules": [
      {
        "from_port": 0,
        "group": {},
        "id": "d87e999b-5b68-44f0-a0ac-5f1801342834",
        "ip_protocol": "icmp",
        "ip_range": {
          "cidr": "2001:558:fc00::/39"
        },
        "parent_group_id": "5f097b2c-5e1a-41a1-b7dc-56c1e182fbb3",
        "to_port": 0
      }
    ],
    "tenant_id": "c212707a4ec34f43ac88e3c1a43ee7b0"
  }
}
```

Now to figure out _why_ the protocol has changed. This API was deprecated 2 years ago - it should not be changing.

**Comment by jtopjian** [comment]
On the topic of the security group test, this is a change to Neutron in master: https://docs.openstack.org/releasenotes/neutron/unreleased.html#upgrade-notes

On the topic of the port binding test, I'm getting the following error:

```
--- FAIL: TestAccNetworkingV2Port_portBinding_update (82.67s)
    testing.go:568: Step 4 error: Check failed: Check 6/6 error: openstack_networking_port_v2.port_1: Attribute 'binding.0.profile' expected "null", got "{}"
```

**Comment by kayrus** [comment]
@jtopjian I forgot that port tests are skipped.

**Comment by jtopjian** [comment]
@kayrus I think we need to put the admin-only restriction back:

```
Error: Error updating OpenStack Neutron Port: Request forbidden: [PUT http://192.168.0.35:9696/v2.0/ports/2642571f-1f92-4f0e-87d3-b66201114b85], error message: {"NeutronError": {"message": "((rule:update_port and rule:update_port:binding:host_id) and rule:update_port:binding:profile) is disallowed by policy", "type": "PolicyNotAuthorized", "detail": ""}}
```

But I'm still unsure of the reason for using a string `"null"`. Is that something that the API will ever send back? If not, is there a reason why a user would ever use a string `"null"`?

**Comment by pkqsun** [comment]
> 
> 
> @pkqsun sorry for the delay, I'll try to finish the PR this week. Please note that usually binary releases are not published right after the fix. However you can always compile the binary directly from the source and I suggested to do this in [#756 (comment)](https://github.com/terraform-providers/terraform-provider-openstack/issues/756#issuecomment-494783342)

Hi @kayrus  @jtopjian ,

Regarding compile a private binary with branch kayrus:omit-binding-profile, could you give me some guide or link how to do that ? 
Thanks in advance.

**Comment by pkqsun** [comment]
@kayrus @jtopjian ,

Any update for this fix ?

**Comment by jtopjian** [comment]
@pkqsun Are you still seeing this problem in the latest version of this plugin?

**Comment by pkqsun** [comment]
> 
> 
> @pkqsun Are you still seeing this problem in the latest version of this plugin?

@jtopjian ,
Sorry, I did not use provider 1.20.0 for testing yet since from https://github.com/terraform-providers/terraform-provider-openstack/blob/master/CHANGELOG.md, I guess this fix is not included.

**Comment by jtopjian** [comment]
@pkqsun There was a vendor/library update which might have implicitly fixed this problem. Can you try the latest version and see if this resolves the problem?

**Comment by pkqsun** [comment]
@jtopjian ,

It still failed with previous error.

> Terraform v0.11.14 (has some Incompatibilities issue, we do not upgrade to 0.12 version currently )
> provider.openstack: version = "~> 1.20"

> 2019-07-24 03:29:49,225 DEBUG: module.vlan-net-02-v4-port01.openstack_networking_port_v2.net-02-v4-port01: Creating...
2019-07-24 03:29:49,225 DEBUG:   admin_state_up:           "" => "true"
2019-07-24 03:29:49,226 DEBUG:   all_fixed_ips.#:          "" => "<computed>"
2019-07-24 03:29:49,226 DEBUG:   all_security_group_ids.#: "" => "<computed>"
2019-07-24 03:29:49,226 DEBUG:   all_tags.#:               "" => "<computed>"
2019-07-24 03:29:49,226 DEBUG:   binding.#:                "" => "1"
2019-07-24 03:29:49,226 DEBUG:   binding.0.vif_details.%:  "" => "<computed>"
2019-07-24 03:29:49,226 DEBUG:   binding.0.vif_type:       "" => "<computed>"
2019-07-24 03:29:49,226 DEBUG:   binding.0.vnic_type:      "" => "direct"
2019-07-24 03:29:49,226 DEBUG:   device_id:                "" => "<computed>"
2019-07-24 03:29:49,226 DEBUG:   device_owner:             "" => "<computed>"
2019-07-24 03:29:49,226 DEBUG:   dns_assignment.#:         "" => "<computed>"
2019-07-24 03:29:49,226 DEBUG:   dns_name:                 "" => "<computed>"
2019-07-24 03:29:49,226 DEBUG:   fixed_ip.#:               "" => "1"
2019-07-24 03:29:49,226 DEBUG:   fixed_ip.0.ip_address:    "" => "10.76.98.82"
2019-07-24 03:29:49,226 DEBUG:   fixed_ip.0.subnet_id:     "" => "316c3c7a-fa99-45d0-ae00-fb31e1372734"
2019-07-24 03:29:49,226 DEBUG:   mac_address:              "" => "<computed>"
2019-07-24 03:29:49,226 DEBUG:   name:                     "" => "eason-net-02-v4-port01"
2019-07-24 03:29:49,226 DEBUG:   network_id:               "" => "7113fc5d-a805-437c-ac2e-442acc6e7b82"
2019-07-24 03:29:49,226 DEBUG:   port_security_enabled:    "" => "<computed>"
2019-07-24 03:29:49,227 DEBUG:   qos_policy_id:            "" => "<computed>"
2019-07-24 03:29:49,227 DEBUG:   region:                   "" => "<computed>"
2019-07-24 03:29:49,227 DEBUG:   tenant_id:                "" => "<computed>"
2019-07-24 03:29:49,449 DEBUG: 
2019-07-24 03:29:49,450 DEBUG: Error: Error applying plan:
2019-07-24 03:29:49,450 DEBUG: 
2019-07-24 03:29:49,450 DEBUG: 1 error occurred:
2019-07-24 03:29:49,450 DEBUG:  * module.vlan-net-02-v4-port01.openstack_networking_port_v2.net-02-v4-port01: 1 error occurred:
2019-07-24 03:29:49,455 DEBUG:  * openstack_networking_port_v2.net-02-v4-port01: Error creating openstack_networking_port_v2: Request forbidden: [POST https://10.4.166.166:13696/v2.0/ports], error message: {"NeutronError": {"message": "(((rule:create_port and rule:create_port:binding:profile) and rule:create_port:fixed_ips) and rule:create_port:binding:vnic_type) on {'binding:host_id': <neutron_lib.constants.Sentinel object at 0x345ffd0>, u'name': u'eason-net-02-v4-port01', 'allowed_address_pairs': <neutron_lib.constants.Sentinel object at 0x345ffd0>, u'admin_state_up': True, u'network_id': u'7113fc5d-a805-437c-ac2e-442acc6e7b82', 'tenant_id': u'498c8a0646e34dccbe2fd598e9059f95', 'extra_dhcp_opts': None, 'mac_address': <neutron_lib.constants.Sentinel object at 0x345ffd0>, u'binding:vnic_type': u'direct', 'device_owner': '', 'qos_policy_id': None, 'device_id': '', 'port_security_enabled': <neutron_lib.constants.Sentinel object at 0x345ffd0>, u'binding:profile': {}, 'project_id': u'498c8a0646e34dccbe2fd598e9059f95', u'fixed_ips': [{u'subnet_id': u'316c3c7a-fa99-45d0-ae00-fb31e1372734', u'ip_address': u'10.76.98.82'}], 'security_groups': <neutron_lib.constants.Sentinel object at 0x345ffd0>, 'description': ''} by {'domain': None, 'project_name': u'v9781_CTO_BCMT', 'tenant_name': u'v9781_CTO_BCMT', 'project_domain': None, 'timestamp': '2019-07-24 03:30:43.381218', 'auth_token': '806a219e5557439795dfac0f3f06d316', 'resource_uuid': None, 'is_admin': False, 'user': u'7aea1ee662874a29b9d0a11ae2303df5', 'tenant': u'498c8a0646e34dccbe2fd598e9059f95', 'read_only': False, 'project_id': u'498c8a0646e34dccbe2fd598e9059f95', 'user_id': u'7aea1ee662874a29b9d0a11ae2303df5', 'show_deleted': False, 'roles': [u'_member_'], 'user_identity': u'7aea1ee662874a29b9d0a11ae2303df5 498c8a0646e34dccbe2fd598e9059f95 - - -', 'tenant_id': u'498c8a0646e34dccbe2fd598e9059f95', 'is_admin_project': True, 'request_id': 'req-ae8e095e-9f1f-462e-9f41-bce1c6e9874f', 'user_domain': None, 'user_name': u'v9781_clivez'} disallowed by policy", "type": "PolicyNotAuthorized", "detail": ""}}
2019-07-24 03:29:49,455 DEBUG:

**Comment by pkqsun** [comment]
update:

After replace provider terraform-provider-openstack_v1.20.0_x4 to a private one that built based on https://github.com/kayrus/terraform-provider-openstack/tree/omit-binding-profile (terraform version still using v0.11.14), it works well for SRIOV port.

**Comment by kayrus** [comment]
recheck

File: openstack/resource_openstack_networking_port_v2.go (hunk 1 above)
Comment (kayrus) [reply openstack/resource_openstack_networking_port_v2.go]: @jtopjian you were right. I removed the `null`.
